keras_nlp/preprocessing/mlm_masker.py

```diff
@@ -0,0 +1,126 @@
+    This class can both be applied in tf.data pipeline as a standalone utility,
+    or used together with `tf.keras.Model` to generate dynamic mask, which is
+    useful for workflows like RoBERTa training.
+
+    Args:
+        vocabulary: A list of strings or a string string filename path. If
+            passing a list, each element of the list should be a single word
+            piece token string. If passing a filename, the file should be a
+            plain text file containing a single word piece token per line.
+        unselectable_tokens: A list of tokens, defaults to None. Tokens in
+            `unselectable_tokens` will not be selected for masking.
+        mask_token: String, defaults to "[MASK]". The mask token.
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        lm_selection_rate,
+        max_selections,
+        unselectable_token_ids=None,
+        mask_token_id=0,
+        mask_token_rate=0.8,
+        random_token_rate=0.1,
+        padding_token=None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.vocabulary_size = vocabulary_size
+        self.unselectable_token_ids = unselectable_token_ids
+        self.lm_selection_rate = lm_selection_rate
+        self.max_selections = max_selections
+        self.mask_token_rate = mask_token_rate
+        self.random_token_rate = random_token_rate
+        self.padding_token = padding_token
+
+        if mask_token_id >= vocabulary_size:
+            raise ValueError(
+                f"Mask token id should be in range [0, vocabulary_size - 1], "
+                f"but received mask_token_id={mask_token_id}."
+            )
+        self.mask_token_id = mask_token_id
+
+    def call(self, inputs):
+        input_is_ragged = isinstance(inputs, tf.RaggedTensor)
+        if not input_is_ragged:
+            if tf.rank(inputs) == 1:
+                # If inputs is of rank 1, we manually add the batch axis.
+                inputs = inputs[tf.newaxis, :]
+            # Convert to RaggedTensor to avoid masking out padded token.
+            inputs = tf.RaggedTensor.from_tensor(
+                inputs,
+                padding=self.padding_token,
+            )
+        random_selector = tf_text.RandomItemSelector(
+            max_selections_per_batch=self.max_selections,
+            selection_rate=self.lm_selection_rate,
+            unselectable_ids=self.unselectable_token_ids,
+        )
+        mask_values_chooser = tf_text.MaskValuesChooser(
+            self.vocabulary_size,
+            self.mask_token_id,
+            mask_token_rate=self.mask_token_rate,
+            random_token_rate=self.random_token_rate,
+        )
+
+        (
+            masked_input_ids,
+            masked_positions,
+            masked_ids,
+        ) = tf_text.mask_language_model(
+            inputs,
+            item_selector=random_selector,
+            mask_values_chooser=mask_values_chooser,
+        )
+
+        if not input_is_ragged:
```

Comment:
This will return dense only for the tokens right, and ragged for positions and ids?

Padded dense tensors will be the default output state for this type of preprocessing, and we are leaving a lot up to user here...
 - convert 2/3 outputs to dense
 - figure out the dense bounding shape
 - create sample_weights where we padded raggeds

I think we might want to rethink the overall flow of this, with dense padded outputs as the main use case we are targeting.

Reply:
discussed offline.